### ISSUE TITLE
[BUG]  Starting with Firefox 84 on Linux (official mozilla.org build), the entire FF stops responding when after 5sec CAD is in "cleanup"

### ISSUE THREAD

**Issue by saahbs** [mrdokenny/Cookie-AutoDelete#993]
**Describe the bug**

Starting with FF 84 (official mozilla.org build) on Linux, CAD is freezing FF when it starts cleaning up after 5 sec. This worked perfectly before FF84. So after 5sec of closing a site, FF is unresponsive for 10+ seconds. I assume this is due to FF changes causing bad interaction with CAD.

Note I'm on a lightly loaded i7-4790K with 32GiB ram and 0 swap usage - ie. not memory pressure issue or high system load.

Disabling CAD on FF84 makes site closing work normal/quickly again.

**To Reproduce**

1. open new tab on www.youtube.com (or any other website)
2. close the youtube tab
3. wait 5 sec (default cookie cleanup)
4. observe that entire FF is unresponsive (can't scroll, can't switch tabs) <--- THE PROBLEM
5. after additional ~10-15sec, FF84 springs back to life, cookies are cleaned, works as expected until next tab or window closure when CAD cleanup kicks in.

**Expected Behavior** 

Upon completing step 2. above, firefox continues to be responsive.

**Screenshots**

There are no visual artifacts associated with this problem.

**Your System Info**

- OS: GNU/Debian amd64 10.7 "Buster" running kernel 4.19.0-9-amd64
- Browser Info: FF Developer Edition (aurora) Mozilla/5.0 (X11; Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0
- CookieAutoDelete Version: 3.5.1

Also, big thank you for creating CAD, a breath of fresh air in today's highly tracked and controlled internet!! +1

**Comment by kennethtran93** [comment]
What is your CAD Settings, and the expression for YouTube if any?

**Comment by saahbs** [comment]
I don't have any youtube specific settings - this behavior happens on every website which triggers cookie cleanup.

Here is a dump of my CAD core settings:

{
  "settings": [
    {
      "name": "activeMode",
      "value": true
    },
    {
      "name": "delayBeforeClean",
      "value": "5"
    },
    {
      "name": "statLogging",
      "value": true
    },
    {
      "name": "showNumOfCookiesInIcon",
      "value": true
    },
    {
      "name": "showNotificationAfterCleanup",
      "value": true
    },
    {
      "name": "cleanCookiesFromOpenTabsOnStartup",
      "value": false
    },
    {
      "name": "contextualIdentities",
      "value": false
    },
    {
      "id": "DEPRECATED - use localStorageCleanup",
      "name": "localstorageCleanup",
      "value": false
    },
    {
      "name": "notificationOnScreen",
      "value": 3
    },
    {
      "name": "domainChangeCleanup",
      "value": false
    },
    {
      "id": "DEPRECATED - use default expressions",
      "name": "greyCleanLocalstorage",
      "value": false
    },
    {
      "id": "DEPRECATED - use default expressions",
      "name": "whiteCleanLocalstorage",
      "value": false
    },
    {
      "name": "enableNewVersionPopup",
      "value": false
    },
    {
      "name": "debugMode",
      "value": false
    },
    {
      "name": "enableGreyListCleanup",
      "value": true
    },
    {
      "name": "keepDefaultIcon",
      "value": false
    },
    {
      "name": "sizePopup",
      "value": 16
    },
    {
      "name": "sizeSetting",
      "value": 16
    },
    {
      "name": "contextMenus",
      "value": true
    },
    {
      "name": "discardedCleanup",
      "value": false
    },
    {
      "name": "cacheCleanup",
      "value": false
    },
    {
      "name": "cleanExpiredCookies",
      "value": false
    },
    {
      "name": "indexedDBCleanup",
      "value": false
    },
    {
      "name": "localStorageCleanup",
      "value": false
    },
    {
      "name": "manualNotifications",
      "value": true
    },
    {
      "name": "pluginDataCleanup",
      "value": false
    },
    {
      "name": "serviceWorkersCleanup",
      "value": false
    }
  ]
}

**Comment by kennethtran93** [comment]
Can you try the alpha build of 3.6.0 and see if it still persists?

I'm currently on the Dev Edition as well, albeit 84.0b8 64-bit Windows 10, and I have yet to notice the freeze during cleanup (tested by scrolling a page until the notification appears). 

Double check the build number on yours?

**Comment by saahbs** [comment]
My exact release is (copied from about:support):

Name 	Firefox
Version 	84.0b8
Build ID 	20201203211213

I disabled xpinstall signature verification and installed v3.6.0-alpha from github.

The issues persists - there is no improvement.

Checking 'Browser Console' also doesn't show anything during/after whenever CAD cleaning is active and blocking.

**Comment by kennethtran93** [comment]
Any chance you can run a profiler to see if we can narrow down the problem? As I've performed the test many times on Windows, I'm not sure how I can start digging into this. 

Did you perhaps make any changes to about:config? 

Are there any other extensions enabled? 

Maybe an irrelevant question, but how many tabs/windows do you have open?

**Comment by saahbs** [comment]
I also tried disabling all add-ons except CAD and the issue persisted.

I also tried creating new Firefox profile, installing CAD, enabling "Enable Automatic Cleaning" and repro-ing. It froze just like with my main profile.

**Comment by saahbs** [comment]
I have some about:config changes, but I can repro on a fresh profile.

I have between two and ten tabs.

I can repro without any other plugins installed/enabled.

I'll try the profiler.

**Comment by saahbs** [comment]
![image](https://user-images.githubusercontent.com/8399931/102173273-82a32300-3e4f-11eb-8d19-0df8bffb41e6.png)

Here is a performance profile of default FF84 start page, fresh profile with only CAD installed. While I was profiling it, I opened yahoo.com and closed it to trigger CAD cleanup. Note the ~discontinuity in the timeline.

If there is another profile tool or approach I can try, please let me know how. ty!

**Comment by kennethtran93** [comment]
I'm not fully aware of how the profiler works, but there may be a view that shows the function being ran and the time it took. JavaScript/webextension related. Maybe the call tree?

From your screenshot it looks like you did a profile for a tab, not the browser itself.

See profiler.firefox.com and see if you can get the webextension area.

**Comment by kennethtran93** [comment]
Call tree may help to see the total times for each call.

**Comment by kennethtran93** [comment]
Not sure if that same recording could help, but there was a 'track' for Webextension that may be useful. In addition when I ran the profiler via the toolbar icon, I got a page that assisted with analyzing the profile record. That page also had screenshots including what the main window screenshots and the notification popup screenshots. In my case I didn't feel any freezing at all.

**Comment by saahbs** [comment]
Allright, I figured it out. From profiles it looked like entire Firefox is blocking. I also noticed that I couldn't switch mouse focus to windows other than hanging firefox. On a hunch I restarted my window manager (awesomewm) and the problem went away. Doing so I also noticed that I'm getting desktop notifications for cookies being removed.

It appears my WM was misbehaving and blocking (and FF timing out) on desktop notification from CAD through Firefox.

Hopefully this note will help somebody in the future running into similar FF desktop notification issues/freezes on U*X.
